### ISSUE TITLE
Allow the classic natural sort on string columns

### ISSUE THREAD

**Issue by SharadKumar** [prisma/prisma#3707]
## Problem
There is often a need to do natural sort on string columns. Machines' sort is not always sensible for humans.
https://blog.codinghorror.com/sorting-for-humans-natural-sort-order/

Prisma doesn't support this currently in `v2.7.1`.

## Suggested solution
Either configured at schema level for the desired column, or by passing a param to `orderBy`.

## Alternatives
Discussed https://github.com/prisma/prisma/discussions/3693

## Additional context

ASC (`12` should be the last one!)
![image](https://user-images.githubusercontent.com/5488144/93547407-5559ec80-f9a8-11ea-9c55-5ac91f7d32a0.png)

DESC (`6` should be the last one!)
![image](https://user-images.githubusercontent.com/5488144/93547499-84705e00-f9a8-11ea-9560-28a11378c9f5.png)

**Comment by meotimdihia** [comment]
Prisma is good, but too limited features.

**Comment by jasonsilberman** [comment]
Wanted to bump this feature request. There are more and more projects that require fractional indexing of some sort, many of which use variable length string keys. Being able to sort these natively in prisma is quite important to us. It would be great if it was even an experimental feature to add sort options that used the engine's natural sort functionality.

**Comment by janpio** [comment]
Helpful here would be to see how the Prisma Client generated query (that already exists) would need to be modified to get the result you are after.

**Comment by meotimdihia** [comment]
it is good if we have a way to use query RAW for `select` or `where`. ( I'm fine with disabling Type in this case).
http://knexjs.org/guide/raw.html

**Comment by janpio** [comment]
Prisma allows raw data access via SQL as well: https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access Not sure how that is connected to what @SharadKumar asked for though.

**Comment by meotimdihia** [comment]
@janpio but $queryRaw doesn't handle plat to nested data. This is what I am using for my website:
I tried to find libs could convert plat to nest data but there is no good library.

```Typescript
async function getChapters(
  prisma,
  chapter,
  checkVol2Chap1
): Promise<chapters[]> {
  let order = Prisma.sql`naturalsort("chap") DESC NULLS LAST, naturalsort("vol") DESC NULLS LAST, naturalsort("title") DESC NULLS LAST`

  if (checkVol2Chap1) {
    order = Prisma.sql`naturalsort("vol") DESC NULLS FIRST, naturalsort("chap") DESC NULLS LAST, naturalsort("title") DESC NULLS LAST`
  }

  const chapters = await prisma.$queryRaw<any>`
    SELECT a.id, a.hid, a.vol, a.title, a.chap, a.group_name, a.lang, chapters_groups.group_id FROM chapters a
    LEFT JOIN chapters_groups ON chapters_groups.chapter_id = a.id
    WHERE status IN ('ok') AND comic_id = ${chapter.comics.id} AND lang = ${chapter.lang}
    ORDER BY ${order} 
  `
  const r = []
  let i = 0
  for (const chapter of chapters) {
    if (chapter.id != r[i]?.id) {
      if (r[i]) {
        i++
      }
      r[i] = Object.assign(chapter, { chapters_groups: [] })
    }

    if (chapter.group_id) r[i].chapters_groups.push(chapter.group_id)
  }

  return r
}
```
I tried to find libs that could convert plat to nest data but there is no good library.
Tried to use Knex but they don't do it too.

**Comment by meotimdihia** [comment]
Today, I find out this answer: https://stackoverflow.com/a/67975489/329424
If you use Postgresql then it can be sort natural by default.

> -- First create a collation with numeric sorting
> CREATE COLLATION numeric (provider = icu, locale = 'en@colNumeric=yes');
> 
> -- Alter table to use the collation
> ALTER TABLE "employees" ALTER COLUMN "em_code" type TEXT COLLATE numeric;

**Comment by jove4015** [comment]
+1 for this, natural sort is important and there are cases where normal sort is also desired, so you may need to alternate between the two.  Setting the column collation forces you to always use the same sort method for the same column, which is less than desirable.

I would promise an alternative implementation strategy though - just provide an orderByRaw method.

I also find myself needing to sort by calculated values often, which is also not supported.  It would probably satisfy any and all of these use cases to simply provide an orderByRaw mechanism that can be added to otherwise normal queries (without having to move the whole thing out to queryRaw).  Not having this ends up meaning we use queryRaw way more often than we really need to, because we need to support custom sorting in a few very limited edge cases, but we don't want to have to write and maintain two different where clause parsers (one to provide the Prisma where object and the other to provide the raw sql) - it's just not worth using the built in query engine.

Really, methods to inject raw sql for portions of the query (ie, like https://github.com/prisma/prisma/issues/5560) (as opposed to replacing the whole query) in general would make prisma more extensible / workaround-able, and satisfy a wide array of use cases without having to have specific development on each use case.   I can understand not providing selectRaw (could be a difficult to make kosher with typescript), but whereRaw, orderByRaw, havingRaw - at a minimum those three should not involve any further type craziness but would open up functionality options.

**Comment by peacechen** [comment]
@jove4015 I like your proposal because it offers a lot of flexibility. I suspect the push back could come from two sources:

- Prisma supports multiple database technologies that have vastly different syntaxes. Postgres and Mongo are very different and don't map well to each other. Even amongst SQL DBs there are significant differences when it comes to their custom SQL extensions.
-  The db engines could fundamentally change how they generate queries. That could impact the results of the orderByRaw injection.

**Comment by jove4015** [comment]
@peacechen Thanks for the feedback.  I understand that Prisma supports a variety of databases - almost all ORMs do.  Prisma already allows you to use many database-specific features; I think it goes without saying that if you use those features, you're increasing your workload to change database systems down the road.  We can all decide for ourselves whether our project requirements allow for this or not but having done this for a while, I know that it's exceedingly rare for most projects to ever change database systems.  I'm not saying it never happens, I'm just saying it's the tail wagging the dog.

Obviously, this is also the case when worrying about db engines changing their syntax.  Theoretically, Mongo or MySQL could do this with any language feature at any time and render Prisma unusable with that version of that database software (along with any other database driver or implementation).  This is a risk any software that interacts with any other 3rd party software package takes.  My update wouldn't do anything to change this risk, it is already present.

I'm hoping we can finally look past the fact that different databases work differently and look towards implementing the features that will help us get the most out of the databases that we have ultimately chosen to use.  Otherwise, we're just reducing all of the databases to the lowest common denominator, making the choice between postgres or mysql or mongo irrelevant because you can only use the features of each that each holds in common with the others.  At that point, Prisma may as well bundle its own database server simply so that it can maintain compatibility with itself and offer a robust feature set - but since I seriously doubt that's on the roadmap, maybe we could just provide full support for the databases we already have.

**Comment by jtkiesel** [comment]
If you need a workaround for this, and your database supports creating custom collation objects, I was able to successfully leverage one to change the default sort behavior for individual columns. I am using PostgreSQL, but I think this should also work with other databases that support creating custom collations. This is the blog post I learned from: https://blog.testdouble.com/posts/2022-12-12-pg-natural-sorting/

TL;DR:
1. Create a collation for numeric/natural order:
```sql
create collation numeric (provider = icu, locale = 'en-u-kn-true');
```
2. Apply that collation to the column you wish to sort numerically/naturally:
```sql
alter table my_table alter column my_column type character varying(255) collate numeric;
```

After following those steps, you can simply use Prisma's currently built-in `orderBy` on that column and it will sort that column the way you desire.

> [!NOTE]
> This doesn't support use cases that require being able to sort by "natural" *and* "machine" order interchangeably, but in my case I only ever needed to sort this particular column numerically/naturally.
